### PR TITLE
Update finufft package/interface

### LINKED ISSUE #286: FI NUFFT Version 2 updates
I was informed FINUFFT v2.0.0 python package was pushed to PyPI.  To take that upgrade requires some code changes towards the new interface, changing package(dependency) name, and a few other tasks.

Similarly, there will be a matching release for cufinufft python package that keeps their interface close to consistent.

These packages should fix some of the quirks from the original implementation, by supporting both single and double precision and avoiding some extra transposes.

### PR DESCRIPTION
**Pull request by garrettwrong**
Closes #286 

Still WIP, just stashing here for now.

### CODE REVIEW

**Comment by garrettwrong** [comment]
Adding some notes:

1. The conversion to gv2 api itself was relatively straightforward and painless for the control `tests/test_nufft.py`
2. However, in practice for the remaining code, now that we have the ability to run in either precision I'm finding that we are really tickling _a lot_ of (dtype) issues in ASPIRE, see #148.  This manifests as passing double precision frequencies along with single precision signal (which itself is sometimes real sometimes complex). Sometimes classes set to `'single'` actually end up passing doubles somehow in some places (but not all of course). To keep things working I had to restore some copies, but I chose to do so as branches which can emit some crude diagnostic messages for the time being.
3. Some of the unit tests relied on old upcasting behavior to repro  (recall finufft v1x did not actually compute in singles for us).
4. The v2.0.0 wheel set is missing wheels for cp36 on OSX and Windows so those tests are not getting past the environ stage.  We can a) have cp36 wheel built for us b) bumpy anaconda to Python 3.8.

**Comment by garrettwrong** [comment]
Rebased.  This should be rebased on top of #296 once that is in `develop`.  Still thinking about windows.

**Comment by garrettwrong** [comment]
I have this rebased locally after the dtype changes. I will wait until I see the windows wheel on PyPI to push that update. Then, assuming everything is passing, I think it would be ready for reviewing.

File: tests/test_nfft.py [hunk 1]
```diff
@@ -382,7 +386,7 @@ def _testAdjointMany(self, backend, dtype, ntransforms=2):
         result = plan.adjoint(batch)
 
         for r in range(ntransforms):
-            self.assertTrue(np.allclose(result[r], self.adjoint_plane))
+            self.assertTrue(np.allclose(result[r], self.adjoint_plane, atol=atol))
```
Comment (garrettwrong) [inline tests/test_nfft.py]: I will change this to utest_tolerance when I fix any other change requests.

**Comment by garrettwrong** [comment]
This has been rebased with the tox/black changes.

File: src/aspire/nufft/finufft.py [hunk 2]
```diff
@@ -1,94 +1,87 @@
...
-
-        self.cast_output = False
-        if self.dtype != np.float64:
-            logger.debug(
-                "This version of finufftpy is hardcoded to doubles internally"
-                "  casting input to doubles, results cast back to singles."
-            )
-            self.cast_output = True
-            self.dtype = np.float64
+        self.dtype_str = dtype_legacy_string(self.dtype)
```
Comment (janden) [inline src/aspire/nufft/finufft.py]: We can't just send a dtype to Finufft? I thought we fixed that.

File: src/aspire/nufft/finufft.py [hunk 3]
```diff
@@ -1,94 +1,87 @@
...
+            n_trans=self.ntransforms,
+            dtype=self.dtype_str,
+        )
 
-        # Get a handle on the appropriate 1d/2d/3d adjoint function in finufftpy
-        self.adjoint_function = getattr(finufftpy, f"nufft{self.dim}d1{manystr}")
+        # FINUFFT expects column major signal data,
+        #   but we don't want to transpose large arrays.
+        #   We instead reverse the references to the points axes.
+        self._transform_plan.setpts(*self.fourier_pts[::-1])
```
Comment (janden) [inline src/aspire/nufft/finufft.py]: Isn't this already handled by the Finufft Python interface?

**Review by janden — CHANGES_REQUESTED** [review]
Some of this stuff doesn't jibe with what I thought the Finufft interface was doing. If we give it `omega_1, omega_2, omega_3` for the `setpts` command, that should correspond to first, second, and third axis in the signal (slowest to fastest, since we're C ordered). Why do we need to do the order reversing here? Are we relying on the opposite order elsewhere in the ASPIRE code?

**Comment by garrettwrong** [comment]
I'll have to confirm which patches are in 2.0.0 and which are in 2.0.1 ... will do.

**Comment by janden** [comment]
It's very possible that I'm missing something here. I don't think 2.0.0 vs. 2.0.1 should make a difference (just pybind11 vs. ctypes). My speculation is that our code may be implicitly using the old ordering (having frequencies go from fastest to slowest), and we just haven't noticed until now.

**Comment by garrettwrong** [comment]
that's possible, I think I have the exact same code in cufinufft now that you mention it. Okay, I'll take a pass at them again, sorry about that!

File: src/aspire/nufft/finufft.py (hunk 2 above)
Comment (garrettwrong) [reply src/aspire/nufft/finufft.py]: It was fixed at some point after I made the initial changeset.  Updated.

File: src/aspire/nufft/finufft.py (hunk 3 above)
Comment (garrettwrong) [reply src/aspire/nufft/finufft.py]: Yes it is. I replaced the ordering of frequencies in any code under test, and removed this here and in the cufinufft class.

**Comment by garrettwrong** [comment]
That was an interesting issue the CI caught... The various wheels for finufft are not available on some older OS platforms, such as the Travis CI default. Wheels in question are only manylinux2010. The default travis CI distros is xenial (16.04), with a tool chain too old to qualify, ie, they will only match against manylinux1.  This caused the pip based pipelines to fail, because the environments could not find/satisfy requirements.  Here I have updated the CI distro to something a little more recent (18.04).

I think in the past, this CI run would have been inside the conda env, and would have spoofed the toolchain as something newer to pickup a recent wheel... tricky.

**Comment by garrettwrong** [comment]
I've updated with the latest `develop`, there were some minor conflicts from the module renaming.

**Comment by janden** [comment]
Ok. Will look through this now. Want to make sure I'm understanding everything with the axis swapping going on.

File: src/aspire/nufft/finufft.py [hunk 4]
```diff
@@ -10,84 +10,73 @@
...
-            self.cast_output = True
-            self.dtype = np.float64
-
         self.complex_dtype = complex_type(self.dtype)
 
-        # TODO: Things get messed up unless we ensure a 'C' ordering here - investigate why
-        self.fourier_pts = np.asarray(
-            np.mod(fourier_pts + np.pi, 2 * np.pi) - np.pi, order="C", dtype=self.dtype
+        self.fourier_pts = np.ascontiguousarray(
+            np.mod(fourier_pts + np.pi, 2 * np.pi) - np.pi
```
Comment (janden) [inline src/aspire/nufft/finufft.py:37]: Ideally, this modulus should be handled by Finufft. If we're getting errors here, we should report it upstream.

**Review by janden — DISMISSED** [review]

**Review by janden — APPROVED** [review]

**Pull request merged**